File: ios/Services/CachedHoleStore.swift
import Foundation

final class CachedHoleStore {
    private var cache: [String: CachedHole] = [:]

    func set(_ hole: CachedHole) {
        cache[hole.holeId] = hole
    }

    func get(holeId: String) -> CachedHole? {
        cache[holeId]
    }

    func clearStale(hours: Double) {
        let cutoff = Date().addingTimeInterval(-hours * 3600)
        cache = cache.filter { _, hole in
            guard let date = ISO8601DateFormatter().date(from: hole.lastSyncedAt) else { return true }
            return date >= cutoff
        }
    }
}

struct CachedHole: Codable {
    let holeId: String
    let pinLat: Double
    let pinLon: Double
    let layups: [LayupTarget]
    let lastSyncedAt: String
    let caddieRecommendation: [String: AnyCodable]?
}

struct LayupTarget: Codable {
    let id: String
    let name: String
    let distanceMeters: Double
    let hazardDistanceMeters: Double?
}

struct AnyCodable: Codable {}

File: shared/cache/CachedHoleStore.ts
export interface LayupTarget {
  id: string;
  name: string;
  lat: number;
  lon: number;
  distanceMeters: number;
  hazardDistanceMeters?: number | null;
}

export interface CachedHole {
  holeId: string;
  pin: { lat: number; lon: number };
  layups: LayupTarget[];
  lastSyncedAt: string;
  caddieRecommendation?: Record<string, unknown>;
}

export class CachedHoleStore {
  private cache = new Map<string, CachedHole>();

  set(hole: CachedHole): void {
    this.cache.set(hole.holeId, hole);
  }

  get(holeId: string): CachedHole | undefined {
    return this.cache.get(holeId);
  }

  clearStale(thresholdHours: number): void {
    const cutoff = Date.now() - thresholdHours * 60 * 60 * 1000;
    for (const [key, value] of this.cache.entries()) {
      const timestamp = Date.parse(value.lastSyncedAt);
      if (!Number.isNaN(timestamp) && timestamp < cutoff) {
        this.cache.delete(key);
      }
    }
  }
}